### ISSUE TITLE
How to fix PLSc bootstrapping issues?

### ISSUE THREAD

**Issue by simvrh** [sem-in-r/seminr#239]
I'm aware that PLSc models can result in errors:

> Note: SEMinR also uses PLSc on PLS models with interactions involving reflective constructs. PLS
models with interactions can be estimated as PLS consistent, but are subject to some bias as per
Becker et al. (2018). It is not uncommon for bootstrapping PLSc models to result in errors due the
calculation of the adjustment.

I have only reflective constructs in my model:

> library(seminr)
> 
> mm <- constructs(
>   reflective('PM',    multi_items('PM', 1:3)),
>   reflective('FoGI',  multi_items('FoGI', 1:3)),
>   reflective('TiISP', multi_items('TiISP', c(1, 3))),
>   reflective('TiSMP', multi_items('TiSMP', 1:3)),
>   reflective('TiG',   multi_items('TiG', 1:3)),
>   reflective('PT',    multi_items('PT', 1:2)),
>   reflective('PC',    multi_items('PC', c(1, 3))),
>   reflective('LoC',   multi_items('LoC', 1:2))
> )
> 
> sm <- relationships(
>   paths(from = c('FoGI', 'TiG', 'TiISP', 'TiSMP'), to = c('PT', 'PC')),
>   paths(from = c('PT', 'PC', 'LoC'), to = 'PM')
> )
> 
> pls.est <- estimate_pls(data.df, mm, sm, inner_weights = path_weighting)
> 
> pls.boot <- bootstrap_model(pls.est, nboot = 500, cores = 2)

No matter what I try to change in the measurement or structural models, I keep getting the same error:

`Error in checkForRemoteErrors(val) (...): 2 nodes produced errors; first error: missing value where TRUE/FALSE needed`

If I change all constructs to formative, then it works like a charm (I guess this switches from PLSc to PLS):

> mm <- constructs(
>   composite('PM',    multi_items('PM', 1:3)),
>   composite('FoGI',  multi_items('FoGI', 1:3)),
>   composite('TiISP', multi_items('TiISP', c(1, 3))),
>   composite('TiSMP', multi_items('TiSMP', 1:3)),
>   composite('TiG',   multi_items('TiG', 1:3)),
>   composite('PT',    multi_items('PT', 1:2)),
>   composite('PC',    multi_items('PC', c(1, 3))),
>   composite('LoC',   multi_items('LoC', 1:2))
> )

Is there any way around these issues, any hints what might be causing them, etc.? They are not happening if I use the `mobi` dataset.

**Comment by soumyaray** [comment]
Thank you @simvrh for bringing this to our attention.  We will try to get to the bottom of this soon and might require more information from you about your model/data.  

In the meantime, if you are familiar with covariance-based structural equation modeling (ala LISREL/Lavaan/MPlus/etc.), you can also try to estimate your reflective model using `estimate_cbsem(...)`. In fact, for a purely reflective model, CBSEM is preferable to PLSc (PLSc merely tries to estimate, post-hoc, what a CBSEM model would have given).  Note that the CBSEM summary report should include path significances without any need to bootstrap.  Do note that CBSEM models come with their own complications, but do let us know if it runs without error.

**Comment by simvrh** [comment]
Thanks a lot @soumyaray for looking into this. Please let me know if you need the full source and data.

`estimate_cbsem` works like a charm. I'm analyzing data from an exploratory study which suggests PLS-SEM may be a more suitable method than CB-SEM (Hair et al., 2017). That is why I'm trying to "migrate" from CB-SEM to PLS-SEM for this particular study.

Hair, J. F. J., Matthews, L. M., Matthews, R. L., & Sarstedt, M. (2017). PLS-SEM or CB-SEM: updated guidelines on which method to use. International Journal of Multivariate Data Analysis, 1(2), 107. https://doi.org/10.1504/IJMDA.2017.087624

**Comment by simvrh** [comment]
I played with `estimate_pls` a bit more. I noticed another potential issue (factor loadings > 1, all VIF < 1.7). Considering the last post on the SmartPLS forum thread below, would it be possible to manually choose PLS / PLSc (for both `estimate_pls` and `bootstrap_model`)?

https://forum.smartpls.com/viewtopic.php?t=3766

I suppose that this might be a request for a new feature (as a quick fix alternative to investigating the issue).

**Comment by NicholasDanks** [comment]
Hi

Glad to see you're getting some utility from our package. 
We provide a fairly strict syntax for estimating SEM models in our package, however, you certainly can select a model that estimates reflectively (mode A) measured constructs and uses regular PLS for the estimation. 
when you select ```reflective()``` type constructs and ```estimate_pls()```, the algorithm automatically applies PLSc. However, if you specify ```composite( ... ,weights = mode_A)``` (the default argument for weights is mode_A so you can also omit this argument) type constructs with ```estimate_pls()``` the algorithm will estimate "reflective" mode A composites with PLS estimation. 

For your model, thus the code you specified above:
```
mm <- constructs(
composite('PM', multi_items('PM', 1:3)),
composite('FoGI', multi_items('FoGI', 1:3)),
composite('TiISP', multi_items('TiISP', c(1, 3))),
composite('TiSMP', multi_items('TiSMP', 1:3)),
composite('TiG', multi_items('TiG', 1:3)),
composite('PT', multi_items('PT', 1:2)),
composite('PC', multi_items('PC', c(1, 3))),
composite('LoC', multi_items('LoC', 1:2))
)
```
will estimate "reflective" mode A composites. For "formative" mode B composites, you can specify the following mm:
```
mm <- constructs(
composite('PM', multi_items('PM', 1:3), weights = mode_B),
composite('FoGI', multi_items('FoGI', 1:3), weights = mode_B),
composite('TiISP', multi_items('TiISP', c(1, 3)), weights = mode_B),
composite('TiSMP', multi_items('TiSMP', 1:3), weights = mode_B),
composite('TiG', multi_items('TiG', 1:3), weights = mode_B),
composite('PT', multi_items('PT', 1:2), weights = mode_B),
composite('PC', multi_items('PC', c(1, 3)), weights = mode_B),
composite('LoC', multi_items('LoC', 1:2), weights = mode_B)
)
```
Hope this helps!

Happy modeling,
Regards,
Nick

**Comment by simvrh** [comment]
Thanks a lot! Not too intuitive but works perfectly!

It gets a bit confusing because of the need to specify the measurement model differently for CB-SEM and PLS-SEM but at least I can live with this. Might be worth including a word or two about this (`composite` can be used to model **both** reflective `weights = mode_A` and formative contructs `weights = mode_B`) in the SEMinR PDF.

PS: Switching to composite also avoids issues with missing rhoC (`Warning in sqrt(rho %*% t(rho)) : NaNs produced` and `Warning in sqrt(rho[i, ]) : NaNs produced`). I didn't open a new issue for this one because of this workaround.

**Comment by soumyaray** [comment]
Greetings @simvrh -- we agree with you on the confusion of terms.  Within SEMinR team, we do not usually refer to mode A/B composites as reflective/formative, though I am not entirely averse to calling them reflective-composite and formative-composite.  Moreover, personally, I want people to think of 'reflective' as being reflective in the common-factor sense. Our thinking on this is always evolving.

As for the inconvenience of having to define everything twice, we've got a rather nice solution: use the `as.reflective()` function to convert your entire measurement model of composites into a reflectives-only model! For example, if you had a measurement model (defined by `constructs(...)`) called `pls_mm` where everything was a composite, then run it as a purely reflective model in cbsem using:

```r
cbsem_model <- estimate_cbsem(as.reflective(pls_mm), sm)
```

In the meantime, it would indeed help to get a copy of your model/data to try out the PLSc estimation. If you are still interested in pursuing the PLSc run, could I ask you to send me your files to soumya.ray@gmail.com ? Everything you send will be kept strictly confidential and destroyed once the issue is closed.

**Comment by NicholasDanks** [comment]
Addressed by issue #242
